src/java.base/share/classes/java/security/PEMDecoder.java

```diff
@@ -0,0 +1,387 @@
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Decodes and returns the specified class for the given PEM stream.  The
+     * class must extend {@link DEREncodable} and be an appropriate class for
+     * the PEM type.
+     *
+     * <p>See {@link PEMDecoder#decode(String, Class)} for details about
+     * {@code tClass}.
+     * <br>See {@link PEMDecoder#decode(InputStream)} for details on using an
+     * {@code InputStream}.
+     *
+     * @param <S> Class type parameter that extends {@code DEREncodable}
+     * @param is an InputStream containing PEM data.
+     * @param tClass the returned object class that implementing
+     *   {@code DEREncodable}.
+     * @return  tClass.
+     * @throws IOException on IO error with the InputStream.
+     * @throws IllegalArgumentException on error in decoding or if the PEM is
+     * unsupported.
+     * @throws ClassCastException if the given class is invalid for the PEM .
+     *
+     */
+    public <S extends DEREncodable> S decode(InputStream is, Class<S> tClass)
+        throws IOException {
+        Objects.requireNonNull(is);
+        Objects.requireNonNull(tClass);
+        PEMRecord pem = Pem.readPEM(is);
+        if (pem == null) {
+            throw new IllegalArgumentException("No PEM data found.");
+        }
+
+        DEREncodable so = decode(pem);
+
+        /*
+         * If the object is a KeyPair, check if the tClass is set to class
+         * specific to a private or public key.  Because PKCS8v2 can be a
+         * KeyPair, it is possible for someone to assume all their PEM private
+         * keys are only PrivateKey and not KeyPair.
+         */
+        if (so instanceof KeyPair kp) {
+            if ((PrivateKey.class).isAssignableFrom(tClass) ||
+                (PKCS8EncodedKeySpec.class).isAssignableFrom(tClass)) {
+                so = kp.getPrivate();
+            }
+            if ((PublicKey.class).isAssignableFrom(tClass) ||
+                (X509EncodedKeySpec.class).isAssignableFrom(tClass)) {
+                so = kp.getPublic();
+            }
+        }
+
+        /*
+         * KeySpec use getKeySpec after the Key has been generated.  Even though
+         * returning a binary encoding after the Base64 decoding is ok when the
+         * user wants PKCS8EncodedKeySpec, generating the key verifies the
+         * binary encoding and allows the KeyFactory to use the provider's
+         * KeySpec()
+         */
+
+        if ((EncodedKeySpec.class).isAssignableFrom(tClass) &&
+            so instanceof Key key) {
+            try {
+                // unchecked suppressed as we know tClass comes from KeySpec
+                // KeyType not relevant here.  We just want KeyFactory
+                if ((PKCS8EncodedKeySpec.class).isAssignableFrom(tClass)) {
+                    getKeyFactory(key.getAlgorithm()).
+                        getKeySpec(key, PKCS8EncodedKeySpec.class);
```

Comment:
Did you forget to assign above to `so`?

Reply:
Ah yes, and that may address a bug found by testing a few days ago.